ethers-solc/src/config.rs

```diff
@@ -547,6 +555,74 @@ pub trait ArtifactOutput {
             })
             .collect()
     }
+
+    /// This returns a maybe versioned artifact path, e.g., either
+    /// `Greeter.sol/Greeter.json` or `Greeter.sol/Greeter.version-string-xxxx.json`
+    /// depending on whether or not `version` is passed in. This also retuns
+    /// the same with the relative source path preserved.
+    /// Defaults to `Greeter.json` if there are any errors.
+    fn versioned_artifact_path(
+        source_path: impl AsRef<Path>,
+        version: Option<impl AsRef<str>>,
+        name: impl AsRef<str>,
+        source_dir: impl AsRef<Path>,
+    ) -> (PathBuf, PathBuf) {
+        let contract_name = source_path
+            .as_ref()
+            .file_name()
+            .map(PathBuf::from)
+            .unwrap_or_else(|| PathBuf::from(name.as_ref()));
+        let art_path = version
+            .map(|ver| Self::output_file_name(format!("{}.{}", name.as_ref(), ver.as_ref())))
+            .unwrap_or_else(|| Self::output_file_name(name));
+        let rel_path = source_path
+            .as_ref()
+            .strip_prefix(source_dir.as_ref())
+            .map(PathBuf::from)
+            .unwrap_or_else(|_| art_path.clone());
+        (rel_path.join(art_path.clone()), contract_name.join(art_path))
+    }
```

Comment:
nit: saves you from redundant `as_ref` calls 
```suggestion
    fn versioned_artifact_path(
        source_path: impl AsRef<Path>,
        version: Option<&str>,
        name: &str,
        source_dir: impl AsRef<Path>,
    ) -> (PathBuf, PathBuf) {
        let contract_name = source_path
            .as_ref()
            .file_name()
            .map(PathBuf::from)
            .unwrap_or_else(|| PathBuf::from(name));
        let art_path = version
            .map(|ver| Self::output_file_name(format!("{}.{}", name, ver)))
            .unwrap_or_else(|| Self::output_file_name(name));
        let rel_path = source_path
            .as_ref()
            .strip_prefix(source_dir)
            .map(PathBuf::from)
            .unwrap_or_else(|_| art_path.clone());
        (rel_path.join(&art_path), contract_name.join(&art_path))
    }
```